### PR TITLE
Feat/devices generate feature parent

### PR DESCRIPTION
**Pull request by InteractiveTimmy**
# Pull Request

## Description

This is a rework of the [draft pull request](https://github.com/webex/webex-js-sdk/pull/1413) to rework `internal-plugin-wdm`.

## Type of Change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Test Coverage

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Generate new tests to accommodate for the new features.
- [x] Updated existing tests to accommodate for the new features.

**Test Configuration**:
* Node/Browser Version v8.16.0
* NPM Version v6.4.1

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/node_modules/@webex/internal-plugin-devices/src/devices.js [hunk 1]
```diff
@@ -12,7 +15,12 @@ const Devices = WebexPlugin.extend({
   extraProperties: 'allow',
 
   children: {
-    // Child FeatureModel will be implemented here in a later pull request.
+    /**
+     * The class object that contains all of the feature collections.
+     *
+     * @type {FeaturesModel}
+     */
+    features: FeaturesModel
```
Comment (InteractiveTimmy) [inline packages/node_modules/@webex/internal-plugin-devices/src/devices.js]: I added the `FeaturesModel` child here.

File: packages/node_modules/@webex/internal-plugin-devices/src/devices.js [hunk 2]
```diff
@@ -1,6 +1,9 @@
 // Internal Dependencies
 import {WebexPlugin} from '@webex/webex-core';
 
+// Local Dependencies
```
Comment (InteractiveTimmy) [inline packages/node_modules/@webex/internal-plugin-devices/src/devices.js]: I updated this comment to be `Local` not `Internal`, as ESLint separates these on save.

File: packages/node_modules/@webex/internal-plugin-devices/src/devices.js [hunk 3]
```diff
@@ -320,7 +328,28 @@ const Devices = WebexPlugin.extend({
...
 
-  initialize() {} // Sets up the plugin as well as its events.
+  /**
+   * Initializer method for the devices plugin.
+   *
+   * @override
+   * @param {Array<any>} args - An array of items to be mapped as properties.
+   * @returns {void}
+   */
+  initialize(...args) {
```
Comment (InteractiveTimmy) [inline packages/node_modules/@webex/internal-plugin-devices/src/devices.js]: I added the FeaturesModel event triggers here, now that features are implemented.

File: packages/node_modules/@webex/internal-plugin-devices/test/integration/spec/wdm-dto.json [hunk 4]
```diff
@@ -0,0 +1,112 @@
+{
```
Comment (InteractiveTimmy) [inline packages/node_modules/@webex/internal-plugin-devices/test/integration/spec/wdm-dto.json]: this file can mostly be ignored, as it's a duplicate of an approved file from a previous PR to allow it's usage in integration tests.

File: packages/node_modules/@webex/internal-plugin-devices/test/unit/spec/devices.js [hunk 5]
```diff
@@ -6,50 +6,50 @@ import MockWebex from '@webex/test-helper-mock-webex';
...
-  let webex;
-  let devices;
-
-  beforeEach('initialize webex with the devices plugin', () => {
-    webex = new MockWebex({
-      children: {
-        devices: Devices
-      }
-    });
+  describe('Devices', () => {
```
Comment (InteractiveTimmy) [inline packages/node_modules/@webex/internal-plugin-devices/test/unit/spec/devices.js]: These changes can be ignored as it was simply fixing an indention style due to an additional `describe()`.

File: packages/node_modules/@webex/internal-plugin-devices/test/unit/spec/features/feature-model.js [hunk 6]
```diff
@@ -3,137 +3,139 @@ import {FeatureModel} from '@webex/internal-plugin-devices';
...
-
-      it('should assign the \'lastModified\' value as a \'Date\'', () => {
-        assert.instanceOf(featureModel.lastModified, Date);
-
-        assert.equal(
-          featureModel.lastModified.toISOString(),
-          featureLM.lastModified
-        );
-      });
+describe('plugin-devices', () => {
```
Comment (InteractiveTimmy) [inline packages/node_modules/@webex/internal-plugin-devices/test/unit/spec/features/feature-model.js]: This file can mostly be ignored as I fixed an indention mismatch due to a missing `describe()`.

File: packages/node_modules/@webex/internal-plugin-devices/src/devices.js [hunk 7]
```diff
@@ -320,7 +328,28 @@ const Devices = WebexPlugin.extend({
...
+    // Prototype the extended class in order to preserve the parent member.
+    Reflect.apply(WebexPlugin.prototype.initialize, this, args);
+
+    // Declare the feature collection group names.
+    const featureCollectionNames = ['developer', 'entitlement', 'user'];
```
Comment (adamweeks) [inline packages/node_modules/@webex/internal-plugin-devices/src/devices.js]: Could we make these constants and exported at the root of the plugin?

File: packages/node_modules/@webex/internal-plugin-devices/src/features/features-model.js [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+   *
+   * @override
+   * @returns {void}
+   */
+  initialize() {
+    // Declare the collection event names.
+    const eventNames = ['change:value', 'add', 'remove'];
+
+    // Declare the collection names.
+    const collectionNames = ['developer', 'entitlement', 'user'];
```
Comment (adamweeks) [inline packages/node_modules/@webex/internal-plugin-devices/src/features/features-model.js]: Could use the constants here

File: packages/node_modules/@webex/internal-plugin-devices/test/integration/spec/devices.js [hunk 9]
```diff
@@ -1,21 +1,59 @@
...
+      devices = webex.internal.devices;
+    });
+
+    // A later pull request will implement this method once this test utilizes
+    // `register()`.
     afterEach('attempt to unregister the device', () => {});
+
     after('remove users', () => testUsers.remove([user]));
 
+    describe('events', () => {
```
Comment (adamweeks) [inline packages/node_modules/@webex/internal-plugin-devices/test/integration/spec/devices.js]: Could these live in unit tests? They aren’t doing any network requests.

File: packages/node_modules/@webex/internal-plugin-devices/test/integration/spec/features/features-model.js [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+  FeatureModel,
+  FeaturesModel
+} from '@webex/internal-plugin-devices';
+import sinon from 'sinon';
+
+import dto from '../wdm-dto';
+
+const collectionNames = Object.keys(dto.features);
+
+describe('plugin-devices', () => {
```
Comment (adamweeks) [inline packages/node_modules/@webex/internal-plugin-devices/test/integration/spec/features/features-model.js]: This one seems like a unit test too

File: packages/node_modules/@webex/internal-plugin-devices/src/devices.js (hunk 7 above)
Comment (InteractiveTimmy) [reply packages/node_modules/@webex/internal-plugin-devices/src/devices.js]: done

File: packages/node_modules/@webex/internal-plugin-devices/test/integration/spec/devices.js (hunk 9 above)
Comment (InteractiveTimmy) [reply packages/node_modules/@webex/internal-plugin-devices/test/integration/spec/devices.js]: moved.

File: packages/node_modules/@webex/internal-plugin-devices/test/integration/spec/features/features-model.js (hunk 10 above)
Comment (InteractiveTimmy) [reply packages/node_modules/@webex/internal-plugin-devices/test/integration/spec/features/features-model.js]: Moved.

File: packages/node_modules/@webex/internal-plugin-devices/src/features/features-model.js (hunk 8 above)
Comment (InteractiveTimmy) [reply packages/node_modules/@webex/internal-plugin-devices/src/features/features-model.js]: done.

File: packages/node_modules/@webex/internal-plugin-devices/src/constants.js [hunk 11]
```diff
@@ -0,0 +1,2 @@
+/* eslint-disable-next-line import/prefer-default-export */
+export const FEATURE_COLLECTION_NAMES = ['developer', 'entitlement', 'user'];
```
Comment (adamweeks) [inline packages/node_modules/@webex/internal-plugin-devices/src/constants.js]: Could each of those be exported as well?

```js
export const FEATURE_COLLECTION_DEVELOPER = 'developer';
export const FEATURE_COLLECTION_ENTITLEMENT = 'entitlement';
export const FEATURE_COLLECTION_USER = 'user';
export const FEATURE_COLLECTION_NAMES = [FEATURE_COLLECTION_DEVELOPER, FEATURE_COLLECTION_ENTITLEMENT, FEATURE_COLLECTION_USER];
```

File: packages/node_modules/@webex/internal-plugin-devices/test/integration/spec/devices.js [hunk 12]
```diff
@@ -27,7 +26,5 @@ describe('plugin-devices', () => {
     describe('#resetLogoutTimer()', () => {});
     describe('#unregister()', () => {});
     describe('#waitForRegistration()', () => {});
-
-    it('temporary pass flag', () => assert.isTrue(true));
```
Comment (adamweeks) [inline packages/node_modules/@webex/internal-plugin-devices/test/integration/spec/devices.js]: Surprised the test library didn't fail due to no tests, but seems to have passed 👍

File: packages/node_modules/@webex/internal-plugin-devices/src/constants.js (hunk 11 above)
Comment (InteractiveTimmy) [reply packages/node_modules/@webex/internal-plugin-devices/src/constants.js]: take care of!

**Pull request merged**